### PR TITLE
Java: Introduce a class of dataflow nodes for the threat modeling.

### PR DESCRIPTION
**Pull request by michaelnebel**
In this PR we introduce a class of data flow nodes called `ThreatModelFlowSource`.
The idea is that these are the exact source nodes we are interested in under a given *threat model*.
For more background information see [EDR](https://github.com/github/code-scanning/blob/main/docs/adrs/0043-threat-modeling.md).

With this implementation it will also be easy to for a query to "opt-in" to *threat models* as defined by the `ThreatModelFlowSource` by updating the data flow configuration.
```codeql
override predicate isSource(DataFlow::Node src) { src instanceof ThreatModelFlowSource }
```

With this change, we can convert most of the data flow related queries that are based on `RemoteFlowSource` to `ThreatModelFlowSource` as this (should) yield the same results as long as the default threat model is used (the default *only* contains *remote*).

### CODE REVIEW

**Comment by michaelnebel** [comment]
DCA looks good!

File: java/ql/lib/semmle/code/java/dataflow/FlowSources.qll [hunk 1]
```diff
@@ -175,14 +206,29 @@ abstract class UserInput extends DataFlow::Node { }
...
+abstract class LocalUserInput extends UserInput, SourceNode {
+  override string getThreatModel() { result = "local" }
+}
 
 /**
  * A node with input from the local environment, such as files, standard in,
  * environment variables, and main method parameters.
  */
-class EnvInput extends LocalUserInput {
+private class EnvInput extends LocalUserInput {
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/dataflow/FlowSources.qll:217]: You're making some classes private, which maybe ought to go through deprecation. It should be simple to use a slightly tweaked private name and then expose a deprecated alias. It seems reasonable to expect that there could be custom queries out there that explicitly reference some of these sources, so they'd need some guidance on how to convert to the new setup.

Comment (michaelnebel) [reply java/ql/lib/semmle/code/java/dataflow/FlowSources.qll]: That is a valid point. I will re-expose a re-implementation of `EnvInput` and deprecate it. Furthermore, also add some hints, how to add implement something that is equivalent. If it turns out that it is really needed by customers that they have this exact class of nodes we can consider removing the deprecation again.

File: java/ql/lib/semmle/code/java/dataflow/FlowSources.qll [hunk 2]
```diff
@@ -191,23 +255,35 @@ class EnvInput extends LocalUserInput {
...
         .(ConstructorCall)
         .getConstructedType()
         .hasQualifiedName("java.io", "FileInputStream")
   }
+
+  override string getThreatModel() { result = "file" }
 }
 
-/** A node with input from a database. */
-class DatabaseInput extends LocalUserInput {
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/dataflow/FlowSources.qll:209]: This should probably have a similar deprecation.

Comment (michaelnebel) [reply java/ql/lib/semmle/code/java/dataflow/FlowSources.qll:209]: Doh!

**Review by aschackmull — APPROVED** [review]
I've not really looked through the tests and their output, but the QL LGTM now.

**Comment by michaelnebel** [comment]
> I've not really looked through the tests and their output, but the QL LGTM now.

Thank you @aschackmull !

**Pull request merged**